### PR TITLE
Multiple touches of a stylus should be considered from the same device

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR fixes https://github.com/flutter/flutter/issues/156223.

**Problem:** The issue above only occurred when using stylus on Web. Multiple touches were treated as different devices: new devices were added and former devices that had the pointer lifted were never removed. This was further caused by the design that Flutter uses JS's `PointerEvent.pointerId` as the device ID, which increases for each touch.

Flutter had to use `PointerEvent.pointerId` because JS doesn't provide a way to distinguish between multiple styluses. This PR fixes this issue by simply supporting only one stylus, since support for multi-stylus can be really hard. How multi-stylus should be supported can be discussed upon such demands in the future, but at least for now, iPad doesn't support multi-stylus and it's not something I can test.

This PR also rewrote some comments that I got confused by while reading the code.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 1]
```diff
@@ -1133,12 +1150,14 @@ class _PointerAdapter extends _BaseAdapter with _WheelEventListenerMixin {
...
-    final ui.PointerDeviceKind kind = _pointerTypeToDeviceKind(event.pointerType!);
-    return kind == ui.PointerDeviceKind.mouse ? _mouseDeviceId :
-        event.pointerId!.toInt();
+    // All mouse pointer events are given `_mouseDeviceId`, including wheel
+    // events, because wheel events might come before any other PointerEvents,
+    // and wheel PointerEvents don't contain pointerIds.
+    return switch(_pointerTypeToDeviceKind(event.pointerType!)) {
+      ui.PointerDeviceKind.mouse => _mouseDeviceId,
+      ui.PointerDeviceKind.stylus => _stylusDeviceId,
+      _ => event.pointerId!.toInt(),
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/pointer_binding.dart:1159]: I guess `_` case covers the multi-tapping with fingers or in the trackpad?

(Should we make this `switch` exhaustive, for clarity?)

**Review by ditman — APPROVED** [review]
LGTM! Small question about the switch returned in `_getPointerId`.

**Comment by ditman** [comment]
> This PR also rewrote some comments that I got confused by while reading the code.

(PS: Thanks for this!)

File: lib/web_ui/lib/src/engine/pointer_binding.dart (hunk 1 above)
Comment (dkwingsmt) [reply lib/web_ui/lib/src/engine/pointer_binding.dart:1159]: The switch case as well as `_` is just a refactor of what was before the PR, except that I added a branch for `stylus`. I don't think it changes how it behaves for trackpad.

Also, AFAIK, trackpad events are handled as wheel events in `_convertWheelEventToPointerData`, which does not call `_getPointerId`, but always use `_mouseDeviceId` instead.

**Comment by dkwingsmt** [comment]
@ditman Thanks for your suggestion of making the switch exhaustive. I've made the change, during which I found out that I didn't include `invertedStylus`. :) 

I've made it clear that `trackpad` isn't reachable in the switch statement.

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 2]
```diff
@@ -1009,7 +1019,15 @@ class _PointerAdapter extends _BaseAdapter with _WheelEventListenerMixin {
       }
     });
 
-    // Why `domWindow` you ask? See this fiddle: https://jsfiddle.net/ditman/7towxaqp
+    // Move event listeners should be added to `_globalTarget` instead of
+    // `_viewTarget`. This is because `_viewTarget` (the root) captures pointers
+    // by default, meaning a pointer that starts within `_viewTarget` continues
+    // sending move events to its listener even when dragged outside.
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/pointer_binding.dart:1025]: I haven't revisited this yet, BUT, there's a way to have the behavior that we needed _globalTarget for from the _viewTarget!!

See this: https://developer.mozilla.org/en-US/docs/Web/API/Element/setPointerCapture

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 3]
```diff
@@ -1133,12 +1151,21 @@ class _PointerAdapter extends _BaseAdapter with _WheelEventListenerMixin {
...
+    // events, because wheel events might come before any other PointerEvents,
+    // and wheel PointerEvents don't contain pointerIds.
+    return switch(_pointerTypeToDeviceKind(event.pointerType!)) {
+      ui.PointerDeviceKind.mouse => _mouseDeviceId,
+
+      ui.PointerDeviceKind.stylus ||
+      ui.PointerDeviceKind.invertedStylus => _stylusDeviceId,
+
+      // Trackpad processing doesn't call this function.
+      ui.PointerDeviceKind.trackpad => throw Exception('Unreachable'),
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/pointer_binding.dart:1164]: Are we sure of this? What happens if I touch with several fingers on a mac touchpad? Does it get converted to a "touch" event?

**Review by ditman — APPROVED** [review]
Small question about multi-touching on a "trackpad" (those shouldn't be "wheel" events, I think?

File: lib/web_ui/lib/src/engine/pointer_binding.dart (hunk 3 above)
Comment (dkwingsmt) [reply lib/web_ui/lib/src/engine/pointer_binding.dart:1164]: I've tested on mac Chrome and multitouching with trackpad works. I think they're still converted to "touch" device. Apps probably won't even correctly process touch events with a "trackpad" device. Also this method is just not called by listeners to touchpad.

I could have changed this to the default, but when I read the code I thought it should be ok, and I'd rather change my assumption upon reports than leaving an unreachable branch here. :)

File: lib/web_ui/lib/src/engine/pointer_binding.dart (hunk 2 above)
Comment (dkwingsmt) [reply lib/web_ui/lib/src/engine/pointer_binding.dart:1025]: Yeah, that's what I thought as well. Maybe you can change it in a future PR :)

Comment (dkwingsmt) [reply lib/web_ui/lib/src/engine/pointer_binding.dart:1025]: I've added a TODO with issue https://github.com/flutter/flutter/issues/157968

File: lib/web_ui/lib/src/engine/pointer_binding.dart (hunk 3 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/pointer_binding.dart:1167]: Thanks for testing this!

**Pull request merged**